website/addons/owncloud/utils.py

```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+from website.oauth.models import ExternalAccount
+from website.security import encrypt, decrypt
+
+
+class ExternalAccountConverter:
+    """
+        OwnCloudProvider from website.addons.owncloud.models hijacks the oauth
+        framework to serialize account credentials for owncloud.
+        ExternalAccountConverter provides a wrapper class for mapping between
+        the two.
+    """
+    def __init__(self, account=None, host=None, username=None, password=None):
+        self.account = account
+        if account:
+            self.host = account.oauth_key
+            self.username = account.display_name
+            self.password = decrypt(account.oauth_secret)
+        elif not account and not host:
+            self.host = None
+            self.username = None
+            self.password = None
+        else:
+            self.account = ExternalAccount(
+                provider='owncloud',
+                provider_name='owncloud',
+                display_name=username,
+                oauth_key=host,
+                oauth_secret=encrypt(password),
+                provider_id=host
```

Comment:
>may result in EA's that violate uniqueness constraints

Especially when, given that `ExternalAccount` is `unique_on(['provider', 'provider_id'])`, you could only have one account linked per ownCloud instance, currently. 

Consider field changes:
* `provider_id` to store `username` 
* depending on whether or not ownCloud has public user profiles associate with any given account, `profile_url` *could* contain `host`
* it follows the standard convention best if what's stored in `oauth_key` is what one would use to actually make a request, so `encrypt(password)`. `oauth_secret`, which is generally unused, could store `host`. Not necessarily a blocker.